### This script contains the following points:
#### Section 1: Importing libraries, defining project path, importing datasets
- (PRIMARY) Dataset "expeditions.csv" as df_exp

#### Section 2: Initial exploration of primary dataframe (df_exp)
- 2.1 Shape and info of df_exp
- 2.2 Creating a subset for variables of interest named df_exp_reduced
- 2.3 Reviewing quantitative variables (using desc function)
- 2.4 Reviewing qualitative variables (using value_counts function)

#### Section 3: Data Cleaning of primary dataset (df_exp_reduced)
- 3.1 Duplicates: Finding and creating new subset to filter out duplicates named df_exp_clean
- 3.2 Inaccurate and Inconsistent Data: Finding and addressing
    - Understanding Categorical O2 Data
- 3.3 Missing Data: Finding and addressing
    - Search for missing values
    - Deriving new column to check O2 categorical variables
    - Updating 105 occurences with no categorical O2 data to have a value of one for 'is_o2_unkwn' variable
    - Updating our df_exp_clean_2 dataframe to have the updated values for 'is_o2_unkwn' (to be held in new column 'is_o2_unkwn_2') - named df_exp_clean_2
    - Creating flag 'o2_check' to check our O2 categorical variables
    - Dropping column 'is_o2_unkwn' since it has been replaced with 'is_o2_unkwn_2' - named df_exp_clean_3
- 3.4 Data Timeliness: Understanding range of years and count of records
- 3.5 Changing column name "had_o2" to "had_o2_unused"
- 3.6 Mixed-Type Columns: None found
- 3.7 Outliers: Identifying outliers

#### Section 4: Exporting dataframe "df_exp_clean_3" as "expeditions_clean.pkl"

## 1. Importing libraries, defining project path, importing data set expedition.csv

In [1]:
# Importing pandas, numpy, and os
import pandas as pd
import numpy as np
import os
import plotly.express as px
import operator

In [2]:
# Defining project folder path
path = r'C:\Users\prena\05-2023 Himalayan Expeditions Analysis'

In [3]:
# Importing expeditions.csv dataset (PRIMARY)
df_exp = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'expeditions.csv'), index_col = False)

In [4]:
pd.set_option('display.max_columns', None)

In [5]:
pd.set_option("display.max_rows", None)

## 2. Initial exploration of primary dataframe (df_exp)

### 2.1 Shape and info of df_exp

In [6]:
df_exp.shape

(10494, 57)

In [7]:
df_exp.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10494 entries, 0 to 10493
Data columns (total 57 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   peak_id            10494 non-null  object
 1   peak_name          10494 non-null  object
 2   nationality        10494 non-null  object
 3   year               10494 non-null  int64 
 4   season             10494 non-null  object
 5   host_cntr          10494 non-null  object
 6   other_cntrs        3508 non-null   object
 7   sponsor            9677 non-null   object
 8   leaders            10470 non-null  object
 9   rte_1_name         10337 non-null  object
 10  rte_2_name         313 non-null    object
 11  rte_3_name         30 non-null     object
 12  rte_4_name         30 non-null     object
 13  team_asc_1         2709 non-null   object
 14  team_asc_2         100 non-null    object
 15  team_asc_3         11 non-null     object
 16  team_asc_4         4 non-null      objec

### 2.2 Creating a subset of variables of interest

In [8]:
# Create subset with variables of interest
df_exp_reduced = df_exp[['peak_id','peak_name','nationality','year','season','bc_arrived','bc_left','summit_day','total_days','exp_result','max_elev_reached','total_mbrs','mbrs_summited','mbrs_deaths','hired_abc','hired_summits','hired_deaths','is_no_hired_abc','is_o2_not_used','is_o2_climbing','is_o2_descent','is_o2_sleeping','is_o2_medical','is_o2_used','is_o2_unkwn','had_o2']]

In [9]:
df_exp_reduced.head()

peak_id   peak_name nationality  year  season  bc_arrived     bc_left  \
0    ACHN      Aichyn       Japan  2015  Autumn         NaN         NaN   
1    ACHN      Aichyn      France  2015  Autumn         NaN         NaN   
2    ACHN      Aichyn      France  2018  Autumn         NaN         NaN   
3    AMAD  Ama Dablam          UK  1958  Autumn         NaN         NaN   
4    AMAD  Ama Dablam          UK  1959  Spring  1959-04-17  1959-05-30   

   summit_day  total_days        exp_result  max_elev_reached  total_mbrs  \
0  2015-09-03           0           Success              6055           5   
1  2015-09-25           0           Success              6055           9   
2         NaN           0     Did not Climb                 0           8   
3         NaN           0  Route Difficulty              6000           6   
4  1959-05-21          43          Accident              6550           6   

   mbrs_summited  mbrs_deaths  hired_abc  hired_summits  hired_deaths  \
0              5            0          0              0             0   
1              4            0          2              2             0   
2              0            0          1              0             0   
3              0            0          2              0             0   
4              0            2          7              0             0   

   is_no_hired_abc  is_o2_not_used  is_o2_climbing  is_o2_descent  \
0                1               1               0              0   
1                0               1               0              0   
2                0               1               0              0   
3                0               1               0              0   
4                0               1               0              0   

   is_o2_sleeping  is_o2_medical  is_o2_used  is_o2_unkwn  had_o2  
0               0              0           0            0       0  
1               0              0           0            0       0  
2               0              0           0            0       0  
3               0              0           0            0       0  
4               0              0           0            0       1

### 2.3 Looking at Quantitative Variables

In [10]:
# Look at descriptive statistics of quantitative variables
df_exp_reduced.describe()

year    total_days  max_elev_reached    total_mbrs  \
count  10494.000000  10494.000000      10494.000000  10494.000000   
mean    2001.531065     20.280541       7106.911187      5.942920   
std       14.755142     17.655621       1763.608343      5.410798   
min     1905.000000      0.000000          0.000000      0.000000   
25%     1994.000000      4.000000       6530.000000      2.000000   
50%     2006.000000     18.000000       7200.000000      5.000000   
75%     2012.000000     32.000000       8188.000000      8.000000   
max     2020.000000    280.000000       8850.000000     99.000000   

       mbrs_summited   mbrs_deaths     hired_abc  hired_summits  hired_deaths  \
count   10494.000000  10494.000000  10494.000000   10494.000000  10494.000000   
mean        1.867257      0.074805      2.787116       1.010101      0.030303   
std         2.894572      0.374835      5.066210       2.437417      0.282998   
min         0.000000      0.000000      0.000000       0.000000      0.000000   
25%         0.000000      0.000000      0.000000       0.000000      0.000000   
50%         1.000000      0.000000      1.000000       0.000000      0.000000   
75%         3.000000      0.000000      3.000000       1.000000      0.000000   
max        39.000000     10.000000     99.000000      48.000000     11.000000   

       is_no_hired_abc  is_o2_not_used  is_o2_climbing  is_o2_descent  \
count     10494.000000    10494.000000    10494.000000   10494.000000   
mean          0.331713        0.706404        0.249190       0.013913   
std           0.470851        0.455431        0.432565       0.117134   
min           0.000000        0.000000        0.000000       0.000000   
25%           0.000000        0.000000        0.000000       0.000000   
50%           0.000000        1.000000        0.000000       0.000000   
75%           1.000000        1.000000        0.000000       0.000000   
max           1.000000        1.000000        1.000000       1.000000   

       is_o2_sleeping  is_o2_medical    is_o2_used   is_o2_unkwn        had_o2  
count    10494.000000   10494.000000  10494.000000  10494.000000  10494.000000  
mean         0.149800       0.030684      0.282542      0.002287      0.086621  
std          0.356892       0.172469      0.450257      0.047770      0.281292  
min          0.000000       0.000000      0.000000      0.000000      0.000000  
25%          0.000000       0.000000      0.000000      0.000000      0.000000  
50%          0.000000       0.000000      0.000000      0.000000      0.000000  
75%          0.000000       0.000000      1.000000      0.000000      0.000000  
max          1.000000       1.000000      1.000000      1.000000      1.000000

### 2.4 Looking at Qualitative Variables

Below, I'm looking through the values that make up each qualitative field, as well as what value is most commonly recorded:

In [11]:
df_exp_reduced['peak_id'].value_counts(dropna = False)

EVER    2155
AMAD    1429
CHOY    1341
MANA     677
LHOT     430
DHA1     389
MAKA     356
BARU     311
PUMO     265
ANN1     242
KANG     184
HIML     162
ANN4      98
PUTH      98
TILI      87
TUKU      62
NUPT      56
LANG      49
MAK2      47
JANU      44
ANN3      37
LSHR      36
ANNS      36
ANN2      35
GAUR      33
KGUR      31
KTEG      28
GANG      28
DORJ      28
TAWO      27
CHAM      26
HIME      26
CHOL      24
KUSU      22
GYAJ      22
YALU      22
GANC      21
SAIP      20
NEMJ      19
BHRI      19
APIM      19
TKPO      19
CTSE      18
GLAC      18
NUMB      16
TASH      16
DHA2      16
SARI      16
DHAM      15
TLNG      15
NILN      15
THAM      15
GYAC      14
LSIS      13
GAN2      13
RATC      13
CHRW      13
KOTA      13
GAUG      13
MANN      12
DHA4      11
KIRA      11
GIMM      11
TENG      11
JANE      11
NEPA      10
PYRM      10
KWAN      10
GAN1      10
LAMJ      10
KYAS      10
PASA      10
KANB      10
PK29      10
LOBE      10
URKM      10
CHOB      10

In [12]:
df_exp_reduced['peak_name'].value_counts(dropna = False)

Everest                      2155
Ama Dablam                   1429
Cho Oyu                      1341
Manaslu                       677
Lhotse                        430
Dhaulagiri I                  389
Makalu                        356
Baruntse                      311
Pumori                        265
Annapurna I                   242
Kangchenjunga                 184
Himlung Himal                 162
Annapurna IV                   98
Putha Hiunchuli                98
Tilicho                        87
Tukuche                        62
Nuptse                         56
Langtang Lirung                49
Makalu II                      47
Jannu                          44
Annapurna III                  37
Lhotse Shar                    36
Annapurna South                36
Annapurna II                   35
Gaurishankar                   33
Kang Guru                      31
Kangtega                       28
Gangapurna                     28
Dorje Lhakpa                   28
Tawoche       

In [13]:
df_exp_reduced['year'].value_counts(dropna = False)

2009    420
2011    417
2012    412
2006    411
2010    406
2013    399
2008    385
2014    369
2019    366
2007    356
2018    353
2017    347
2016    331
2005    329
2004    297
2003    283
2015    283
2000    256
2002    246
2001    222
1998    218
1999    199
1997    193
1996    187
1991    176
1989    155
1990    147
1994    146
1995    141
1993    139
1992    138
1984    134
1988    133
1987    128
1986    121
1985    113
1983    110
1982     99
1981     95
1980     83
1979     67
1978     56
1974     39
1971     36
1955     35
1953     33
1970     32
1954     31
1975     30
1973     29
1972     29
1964     26
1963     24
1962     23
1977     21
1960     21
1969     20
1976     20
1965     17
1959     15
1952     13
1961     13
1949     12
1958     11
1956      9
1936      7
1950      7
1951      7
1957      6
1930      6
1937      6
1967      6
1935      5
1939      5
1966      3
1910      3
1920      3
2020      3
1968      3
1947      2
1934      2
1909      2
1925      2
1929

In [14]:
df_exp_reduced['nationality'].value_counts(dropna = False)

USA                1167
Japan              1040
UK                  930
France              797
Spain               707
Italy               543
Switzerland         450
Germany             436
S Korea             430
Austria             356
Nepal               254
Australia           246
Poland              237
New Zealand         233
India               225
Canada              222
Russia              208
China               197
Czech Republic      135
Netherlands         127
W Germany           107
Norway               93
Slovenia             93
Iran                 80
Sweden               79
Belgium              74
Ukraine              57
Romania              54
Yugoslavia           53
Mexico               53
Denmark              49
Slovakia             41
Argentina            39
Chile                39
Brazil               36
Ireland              31
Ecuador              31
Czechoslovakia       30
Singapore            30
S Africa             29
Greece               29
Bulgaria        

In [15]:
df_exp_reduced['season'].value_counts(dropna = False).sort_index()

All          2
Autumn    5209
Spring    4860
Summer     106
Winter     317
Name: season, dtype: int64

In [16]:
df_exp_reduced['exp_result'].value_counts(dropna = False)

Success                       5692
Bad Weather                   1308
Bad Conditions                1109
Illness, AMS                   465
Route Difficulty               439
Other                          316
Accident                       296
Did not Climb                  233
Last of Supplies               220
Success (Subpeak, ForeSmt)     121
Unknown                        106
Lack of Time                    92
Not to Reach BC                 64
Success (Claimed)               21
Attempt Rumored                 12
Name: exp_result, dtype: int64

In [17]:
df_exp_reduced['is_no_hired_abc'].value_counts(dropna = False)

0    7013
1    3481
Name: is_no_hired_abc, dtype: int64

Note: For information about qualitative, categorical oxygen data, see below in the section 3.2 (Inaccurate or Inconsistent data).

## 3. Data Cleaning of primary dataset (df_exp_reduced) 

### 3.1 Duplicates

In [18]:
# Search for full duplicates
df_full_dups = df_exp_reduced[df_exp_reduced.duplicated()]

In [19]:
df_full_dups

peak_id peak_name nationality  year  season  bc_arrived     bc_left  \
3673    CHOY   Cho Oyu       Japan  2016  Autumn  2016-09-11  2016-10-04   
4820    EVER   Everest         USA  1999  Spring  1999-04-11  1999-05-29   
6104    EVER   Everest       Nepal  2015  Spring         NaN         NaN   
6150    EVER   Everest        Iran  2016  Spring  2016-04-18  2016-05-24   
7720    LHOT    Lhotse         USA  2000  Spring         NaN         NaN   

      summit_day  total_days     exp_result  max_elev_reached  total_mbrs  \
3673  2016-09-30          23        Success              8188           4   
4820  1999-05-26          48        Success              8850           1   
6104         NaN           0          Other                 0           1   
6150  2016-05-19          36        Success              8850           1   
7720         NaN           0  Did not Climb                 0           4   

      mbrs_summited  mbrs_deaths  hired_abc  hired_summits  hired_deaths  \
3673              4            0          3              3             0   
4820              1            0          1              1             0   
6104              0            0          0              0             0   
6150              1            0          1              1             0   
7720              0            0          0              0             0   

      is_no_hired_abc  is_o2_not_used  is_o2_climbing  is_o2_descent  \
3673                0               0               1              0   
4820                0               0               1              0   
6104                0               1               0              0   
6150                0               0               1              0   
7720                1               1               0              0   

      is_o2_sleeping  is_o2_medical  is_o2_used  is_o2_unkwn  had_o2  
3673               0              0           1            0       0  
4820               0              0           1            0       0  
6104               0              0           0            0       0  
6150               1              0           1            0       0  
7720               0              0           0            0       0

In [20]:
# Check shape of data frame before
df_exp_reduced.shape

(10494, 26)

In [21]:
# Create a new subset for df_exp_reduced that excludes duplicates
df_exp_clean = df_exp_reduced.drop_duplicates()

In [22]:
# Check shape of data frame after excluding duplicates
df_exp_clean.shape

(10489, 26)

There were 5 full duplicates found (stored in "df_full_dups.") I created a new subset to exclude these called "df_exp_clean." Our previous data frame had 10,494 records, and our new data frame excluding these 5 records has 10,489 records. 

### 3.2 Inaccurate or Inconsistent data

#### [How do the categorical O2 columns relate to each other?]
Since we're interested in whether or not oxygen was used on the expedition, I want to pay special attention to our 8 categorical O2 variables. Specifically, each record should fall into one of the following categories:
- O2 is not used (variables 'is_o2_not_used' or 'had_o2')
- O2 is used (variables 'is_o2_climbing', 'is_o2_descent', 'is_o2_sleeping', 'is_o2_medical', or 'is_o2_used')
- If the record doesn't have a 1 value for any of the variables in either of these categories, I would expect the presence of oxygen to be unknown and there should be a value of 1 for the variable 'is_o2_unknown.'

Below, I look at each O2 categorical variable to make sure that there are no contradicting records that indicate O2 is not used and also used on an expedition.

##### VARIABLE 'is_o2_unkwn' 
- Definition: It is unknown as to if oxygen was brought on the expedition.

First, I want to look at the count of records where the possession of O2 is unknown.

In [23]:
# Check frequency for column 'is_o2_unknown'
df_exp_clean['is_o2_unkwn'].value_counts(dropna = False).sort_index()

0    10465
1       24
Name: is_o2_unkwn, dtype: int64

This tells us that there are 24 expeditions where it was unknown whether oxygen was brought our not. I'm going to create a subset where 'is_o2_unkwn' = 1 (in other words, these are the number of records where it is unknown if there was oxygen brought).

In [24]:
# Create subset for 'is_o2_unkwn' = 1
df_o2_unkwn = df_exp_clean[df_exp_clean['is_o2_unkwn']==1]

In [25]:
df_o2_unkwn.head()

peak_id   peak_name  nationality  year  season  bc_arrived     bc_left  \
1087    AMAD  Ama Dablam       Norway  2014  Autumn  2014-10-28  2014-11-06   
1096    AMAD  Ama Dablam           UK  2014  Autumn  2014-11-09  2014-11-17   
1370    AMAD  Ama Dablam        Nepal  2019  Autumn         NaN         NaN   
1398    AMAD  Ama Dablam  Netherlands  2019  Autumn  2019-10-19  2019-11-06   
2215    BARU    Baruntse           UK  2014  Autumn  2014-10-26  2014-11-06   

      summit_day  total_days        exp_result  max_elev_reached  total_mbrs  \
1087  2014-11-03           9    Bad Conditions              6000          10   
1096  2014-11-14           8           Success              6814           2   
1370  2019-10-10           0           Success              6814           0   
1398  2019-10-29          18           Success              6814          10   
2215  2014-11-04          11  Route Difficulty              6995           2   

      mbrs_summited  mbrs_deaths  hired_abc  hired_summits  hired_deaths  \
1087              0            0          5              0             0   
1096              2            0          0              0             0   
1370              0            0          3              3             0   
1398              4            0          8              4             0   
2215              0            0          1              0             0   

      is_no_hired_abc  is_o2_not_used  is_o2_climbing  is_o2_descent  \
1087                0               1               0              0   
1096                1               1               0              0   
1370                0               1               0              0   
1398                0               1               0              0   
2215                0               1               0              0   

      is_o2_sleeping  is_o2_medical  is_o2_used  is_o2_unkwn  had_o2  
1087               0              0           0            1       0  
1096               0              0           0            1       0  
1370               0              0           0            1       0  
1398               0              0           0            1       0  
2215               0              0           0            1       0

##### Comparing "is_o2_unknown" positve to variables that indicate O2 is NOT present:
For this subset, I'm interested as to if there are records that can both have O2 listed as unknown and listed as NOT used.

In [26]:
df_o2_unkwn[df_o2_unkwn['had_o2']==1].shape

(0, 26)

In [27]:
df_o2_unkwn[df_o2_unkwn['is_o2_not_used']==1].shape

(13, 26)

##### Comparing "is_o2_unknown" positve to variables that indicate O2 IS present:
For this subset, I would expect there to be no '1' values for fields that indicate that there is oxygen present. I'm looking for a shape of (0,23) for each of the following variables which will tell me that there are no records where both "is_o2_unknwn" and a variable indicating o2 is present exists:

In [28]:
# Checking variable 'is_o2_used'
df_o2_unkwn[df_o2_unkwn['is_o2_used']==1].shape

(0, 26)

In [29]:
# Checking variable 'is_o2_climbing'
df_o2_unkwn[df_o2_unkwn['is_o2_climbing']==1].shape

(0, 26)

In [30]:
# Checking variable 'is_o2_descent'
df_o2_unkwn[df_o2_unkwn['is_o2_descent']==1].shape

(0, 26)

In [31]:
# Checking variable 'is o2_sleeping'
df_o2_unkwn[df_o2_unkwn['is_o2_sleeping']==1].shape

(0, 26)

In [32]:
# Checking variable 'is_o2_medical'
df_o2_unkwn[df_o2_unkwn['is_o2_medical']==1].shape

(0, 26)

##### VARIABLE 'is_o2_not_used'
- Definition: Oxygen is not used on this expedition.

I want to look at the count of records where oxygen is NOT used.

In [33]:
# Check frequency for column 'is_o2_not_used'
df_exp_clean['is_o2_not_used'].value_counts(dropna = False).sort_index()

0    3078
1    7411
Name: is_o2_not_used, dtype: int64

This tells us that there are 7413 expeditions where oxygen was not used. 

In [34]:
# Create subset for is _o2_not_used = 1
df_o2_not_used = df_exp_clean[df_exp_clean['is_o2_not_used']==1]

In [35]:
df_o2_not_used.head()

peak_id   peak_name nationality  year  season  bc_arrived     bc_left  \
0    ACHN      Aichyn       Japan  2015  Autumn         NaN         NaN   
1    ACHN      Aichyn      France  2015  Autumn         NaN         NaN   
2    ACHN      Aichyn      France  2018  Autumn         NaN         NaN   
3    AMAD  Ama Dablam          UK  1958  Autumn         NaN         NaN   
4    AMAD  Ama Dablam          UK  1959  Spring  1959-04-17  1959-05-30   

   summit_day  total_days        exp_result  max_elev_reached  total_mbrs  \
0  2015-09-03           0           Success              6055           5   
1  2015-09-25           0           Success              6055           9   
2         NaN           0     Did not Climb                 0           8   
3         NaN           0  Route Difficulty              6000           6   
4  1959-05-21          43          Accident              6550           6   

   mbrs_summited  mbrs_deaths  hired_abc  hired_summits  hired_deaths  \
0              5            0          0              0             0   
1              4            0          2              2             0   
2              0            0          1              0             0   
3              0            0          2              0             0   
4              0            2          7              0             0   

   is_no_hired_abc  is_o2_not_used  is_o2_climbing  is_o2_descent  \
0                1               1               0              0   
1                0               1               0              0   
2                0               1               0              0   
3                0               1               0              0   
4                0               1               0              0   

   is_o2_sleeping  is_o2_medical  is_o2_used  is_o2_unkwn  had_o2  
0               0              0           0            0       0  
1               0              0           0            0       0  
2               0              0           0            0       0  
3               0              0           0            0       0  
4               0              0           0            0       1

##### Comparing "is_o2_not_used" positve to other variables that indicate O2 is NOT present:
For this subset, I'm interested as to if there are records that can both have O2 listed as unknown and listed as NOT used.
Remember: We know 13 of these expeditions were also listed as being unknown as to if oxygen was in the expedition's possession. I'm going to create a subset where 'is_o2_not_used' = 1 (in other words, these are the number of records where it was recorded that oxygen was NOT used).

In [36]:
df_o2_not_used[df_o2_not_used['had_o2']==1].shape

(909, 26)

The 'had_o2' variable tells us that there were 909 expeditions that brought oxygen but did NOT use it (this is out of the 7413 expeditions that did not use oxygen).

##### Comparing "is_o2_not_used" positve to variables that indicate O2 IS present:
For this subset, I would expect there to be no '1' values for fields that indicate that there is oxygen present. I'm looking for a shape of (0,23) for each of the following variables which will tell me that there are no records where both "is_o2_unknwn" and a variable indicating o2 is present exists:

In [37]:
df_o2_not_used[df_o2_not_used['is_o2_used']==1].shape

(0, 26)

In [38]:
df_o2_not_used[df_o2_not_used['is_o2_sleeping']==1].shape

(0, 26)

In [39]:
df_o2_not_used[df_o2_not_used['is_o2_medical']==1].shape

(0, 26)

In [40]:
df_o2_not_used[df_o2_not_used['is_o2_descent']==1].shape

(0, 26)

In [41]:
df_o2_not_used[df_o2_not_used['is_o2_climbing']==1].shape

(0, 26)

##### VARIABLE 'had_o2'
- Definition: Oxygen was brought on the expedition but NOT used. (Note: this name is a bit confusing and is changed to 'had_o2_unused' further down in this script for clarification purposes).

In [42]:
# Check frequency for column 'had_o2'
df_exp_clean['had_o2'].value_counts(dropna = False).sort_index()

0    9580
1     909
Name: had_o2, dtype: int64

In [43]:
# Create subset for had_o2 = 1. 
df_had_o2 = df_exp_clean[df_exp_clean['had_o2']==1]

In [44]:
df_had_o2.head()

peak_id   peak_name nationality  year  season  bc_arrived     bc_left  \
4     AMAD  Ama Dablam          UK  1959  Spring  1959-04-17  1959-05-30   
6     AMAD  Ama Dablam      France  1978  Autumn  1978-10-01  1978-10-23   
7     AMAD  Ama Dablam         USA  1979  Spring  1979-03-29  1979-05-02   
8     AMAD  Ama Dablam      France  1979  Autumn  1979-09-25  1979-10-25   
14    AMAD  Ama Dablam       Spain  1981  Spring  1981-03-22         NaN   

    summit_day  total_days      exp_result  max_elev_reached  total_mbrs  \
4   1959-05-21          43        Accident              6550           6   
6   1978-10-20          22  Bad Conditions              6000           8   
7   1979-04-22          34         Success              6814          16   
8   1979-10-21          30         Success              6814          23   
14  1981-04-28           0         Success              6814           9   

    mbrs_summited  mbrs_deaths  hired_abc  hired_summits  hired_deaths  \
4               0            2          7              0             0   
6               0            0          5              0             0   
7               9            0          9              1             0   
8              13            0          5              3             0   
14              7            0          4              3             0   

    is_no_hired_abc  is_o2_not_used  is_o2_climbing  is_o2_descent  \
4                 0               1               0              0   
6                 0               1               0              0   
7                 0               1               0              0   
8                 0               1               0              0   
14                0               1               0              0   

    is_o2_sleeping  is_o2_medical  is_o2_used  is_o2_unkwn  had_o2  
4                0              0           0            0       1  
6                0              0           0            0       1  
7                0              0           0            0       1  
8                0              0           0            0       1  
14               0              0           0            0       1

##### Comparing "had_o2" positve to other variables that indicate O2 is NOT present:
For this subset, I know from above that there are no "had_o2" records that are also "is_o2_unknwn." When looking at the variable "is_o2_not_used", there were 909 records that had "is_o2_not_used" and "had_o2" both checked. I assume that the 909 records we see in this subset are the same we found before in the "is_o2_not_used" subset, but I'll double check here: 

In [45]:
df_had_o2[df_had_o2['is_o2_not_used']==1].shape

(909, 26)

##### Comparing "had_o2" positve to variables that indicate O2 IS present:
For this subset, I would expect there to be no '1' values for fields that indicate that there is oxygen present. I'm looking for a shape of (0,23) for each of the following variables which will tell me that there are no records where both "had_o2" and a variable indicating o2 is present exists:

In [46]:
df_had_o2[df_had_o2['is_o2_used']==1].shape

(0, 26)

In [47]:
df_had_o2[df_had_o2['is_o2_sleeping']==1].shape

(0, 26)

In [48]:
df_had_o2[df_had_o2['is_o2_medical']==1].shape

(0, 26)

In [49]:
df_had_o2[df_had_o2['is_o2_descent']==1].shape

(0, 26)

In [50]:
df_had_o2[df_had_o2['is_o2_climbing']==1].shape

(0, 26)

##### VARIABLE 'is_o2_used'
- Definition: This variable indicates that oxygen was used on the expedition. What I want to explore is if this is an umbrella category (for example, if one of the other 4 O2 variables that indicate oxygen is present is positive, will this variable also be positive?)

In [51]:
# Check frequency for column 'is_o2_used'
df_exp_clean['is_o2_used'].value_counts(dropna = False).sort_index()

0    7527
1    2962
Name: is_o2_used, dtype: int64

This tells us that there are 2960 records where oxygen was used. I now want to create a subset of 'is_o2_used' = 1 to look into this deeper:

In [52]:
# Create subset where 'is_o2_used' = 1
df_is_o2_used = df_exp_clean[df_exp_clean['is_o2_used']==1]

In [53]:
df_is_o2_used.head()

peak_id   peak_name nationality  year  season  bc_arrived     bc_left  \
251    AMAD  Ama Dablam     Germany  1999  Autumn  1999-09-20  1999-10-17   
280    AMAD  Ama Dablam         USA  2000  Spring  2000-04-01  2000-04-18   
308    AMAD  Ama Dablam          UK  2000  Winter  2000-11-26  2000-12-09   
440    AMAD  Ama Dablam         USA  2004  Autumn  2004-11-18  2004-11-28   
446    AMAD  Ama Dablam         USA  2004  Autumn  2004-10-23  2004-11-06   

     summit_day  total_days    exp_result  max_elev_reached  total_mbrs  \
251  1999-10-14          27  Illness, AMS              6300           7   
280  2000-04-14          17   Bad Weather              6280           8   
308  2000-12-07          13       Success              6814           8   
440  2004-11-26          10       Success              6814           6   
446  2004-11-02          14       Success              6814          12   

     mbrs_summited  mbrs_deaths  hired_abc  hired_summits  hired_deaths  \
251              0            0          2              0             0   
280              0            0          3              0             0   
308              2            0          2              1             0   
440              1            0          2              1             0   
446              8            0          4              4             0   

     is_no_hired_abc  is_o2_not_used  is_o2_climbing  is_o2_descent  \
251                0               0               0              1   
280                0               0               0              0   
308                0               0               0              0   
440                0               0               0              0   
446                0               0               0              0   

     is_o2_sleeping  is_o2_medical  is_o2_used  is_o2_unkwn  had_o2  
251               0              0           1            0       0  
280               0              1           1            0       0  
308               0              1           1            0       0  
440               0              1           1            0       0  
446               0              1           1            0       0

##### Comparing "is_o2_used" positve to variables that indicate O2 is NOT present:
We know from above that our O2 variables that indicate oxygen is NOT present that 'is_o2_used' is not positive for any of these variables.

##### Comparing "had_o2" positve to variables that indicate O2 IS present:
For this subset, I want to see how many records there are that have both 'is_o2_used' in addition to another variable that indicates oxygen was used.

In [54]:
# Creating new table to see if the number of records of o2-present variables match that of the 'is_o2_used' variable)
df_o2_used = df_is_o2_used[['is_o2_climbing','is_o2_descent','is_o2_sleeping','is_o2_medical']]

In [55]:
# Create column to sum all variables
df_o2_used_2 = df_o2_used.sum(axis=1)

In [56]:
df_o2_used_2

251      1
280      1
308      1
440      1
446      1
610      1
634      1
734      1
776      1
800      1
1014     1
1047     2
1049     1
1113     1
1156     1
1236     2
1260     1
1279     2
1284     1
1286     3
1290     1
1292     1
1359     1
1403     1
1417     1
1423     1
1435     1
1448     2
1449     2
1451     1
1454     2
1456     1
1457     1
1460     1
1480     2
1482     2
1525     1
1529     2
1535     1
1542     1
1562     1
1573     1
1576     1
1612     2
1616     1
1621     1
1624     3
1637     2
1639     1
1650     3
1655     1
1657     1
1658     1
1668     1
1669     1
1670     1
1680     1
1682     2
1683     2
1684     1
1689     2
1744     1
1789     2
1794     1
1804     1
1810     1
1839     1
1849     1
1857     2
1860     1
1868     1
1883     2
1920     1
1946     1
1947     1
1960     1
2000     1
2027     2
2071     1
2079     1
2104     1
2114     1
2115     1
2119     1
2156     1
2192     1
2232     1
2257     1
2390     1
2400     1
2409     2

In [57]:
# using sum() + generator expression
# to get count of elements matching condition
# check for records greater than 0
res = sum(1 for i in df_o2_used_2 if i > 0)
 
# printing result
print ("The number of elements greater than 0: " + str(res))

The number of elements greater than 0: 2946


In [58]:
# Check frequency for column 'is_o2_used'
df_exp_clean['is_o2_used'].value_counts(dropna = False).sort_index()

0    7527
1    2962
Name: is_o2_used, dtype: int64

Our list shows us that there are 2944 records that show that oxygen was used for either medical, sleeping, climbing, or descent purposes AND the 'is_o2_used' variable is positive. On the other hand, there are a total of 2960 records where 'is_o2_used' is recorded. This tells us that 'is_o2_used' is used as an umbrella variable, and there are 16 additional cases where o2 was used but it was not specified if it was used for medical, sleeping, climbing, or descent.

Now that we know that the 5 o2 variables that indicate oxygen was used on the expedition caught under the umbrella of 'is_o2_used', I want to know the frequencies of each of these remaining O2 variables. 

##### VARIABLE 'is_o2_medical'

In [59]:
# Check frequency for column 'is_o2_medical'
df_exp_clean['is_o2_medical'].value_counts(dropna = False).sort_index()

0    10167
1      322
Name: is_o2_medical, dtype: int64

##### VARIABLE 'is_o2_climbing'

In [60]:
# Check frequency for column 'is_o2_climbing'
df_exp_clean['is_o2_climbing'].value_counts(dropna = False).sort_index()

0    7877
1    2612
Name: is_o2_climbing, dtype: int64

##### VARIABLE 'is_o2_descent'

In [61]:
# Check frequency for column 'is_o2_descent'
df_exp_clean['is_o2_descent'].value_counts(dropna = False).sort_index()

0    10343
1      146
Name: is_o2_descent, dtype: int64

##### VARIABLE 'is_o2_sleeping'

In [62]:
# Check frequency for column 'is_o2_sleeping'
df_exp_clean['is_o2_sleeping'].value_counts(dropna = False).sort_index()

0    8918
1    1571
Name: is_o2_sleeping, dtype: int64

Based on these counts not adding up to an even 2,960, we can assume that expeditions were recorded to use oxygen in multiple ways. (We can also see this above in our 'df_o2_used_2' list were the value is greater than 1). 

In [63]:
# Check frequency for column 'is_o2_used'
df_o2_used_2.value_counts(dropna = False).sort_index()

0      16
1    1350
2    1501
3      81
4      14
dtype: int64

### 3.3 Missing data

#### Search for missing values

In [64]:
# Search for missing values
df_exp_clean.isnull().sum()

peak_id                0
peak_name              0
nationality            0
year                   0
season                 0
bc_arrived          1147
bc_left             2437
summit_day           660
total_days             0
exp_result             0
max_elev_reached       0
total_mbrs             0
mbrs_summited          0
mbrs_deaths            0
hired_abc              0
hired_summits          0
hired_deaths           0
is_no_hired_abc        0
is_o2_not_used         0
is_o2_climbing         0
is_o2_descent          0
is_o2_sleeping         0
is_o2_medical          0
is_o2_used             0
is_o2_unkwn            0
had_o2                 0
dtype: int64

This tells us that there are missing values for bc_arrived, bc_left, and Summit_day. I would like to leave these missing values in for now, as the records still have valuable information, otherwise.

Since we're interested in the presence of 02, I want to pay special attention to our 8 categorical O2 variables. Specifically, each record should fall into one of the following categories:
- O2 is not used (variables 'is_o2_not_used' or 'had_o2')
- O2 is used (variables 'is_o2_climbing', 'is_o2_descent', 'is_o2_sleeping', 'is_o2_medical', or 'is_o2_used')
- If the record doesn't have a 1 value for any of the variables in either of these categories, I would expect the presence of oxygen to be unknown and there should be a value of 1 for the variable 'is_o2_unknown.'

#### Deriving new column to check O2 categorical variables

In [65]:
# Creating new table to see if there are records where the sum of all categorical o2 variables = 0)
df_o2_check = df_exp_clean[['is_o2_not_used','is_o2_climbing','is_o2_descent','is_o2_sleeping','is_o2_medical','is_o2_used','is_o2_unkwn','had_o2']]

In [66]:
df_o2_check.head()

is_o2_not_used  is_o2_climbing  is_o2_descent  is_o2_sleeping  \
0               1               0              0               0   
1               1               0              0               0   
2               1               0              0               0   
3               1               0              0               0   
4               1               0              0               0   

   is_o2_medical  is_o2_used  is_o2_unkwn  had_o2  
0              0           0            0       0  
1              0           0            0       0  
2              0           0            0       0  
3              0           0            0       0  
4              0           0            0       1

In [67]:
# Create list to find 0 values
df_o2_check_2 = df_o2_check.sum(axis=1)

In [68]:
df_o2_check_2

0        1
1        1
2        1
3        1
4        2
5        1
6        2
7        2
8        2
9        1
10       1
11       1
12       1
13       1
14       2
15       1
16       1
17       1
18       1
19       1
20       1
21       1
22       1
23       1
24       1
25       1
26       1
27       1
28       1
29       1
30       1
31       1
32       1
33       1
34       1
35       1
36       1
37       1
38       1
39       1
40       1
41       1
42       1
43       1
44       1
45       1
46       1
47       1
48       1
49       1
50       1
51       1
52       1
53       1
54       1
55       1
56       1
57       1
58       1
59       1
60       1
61       1
62       1
63       1
64       1
65       1
66       1
67       1
68       1
69       1
70       1
71       1
72       1
73       1
74       1
75       1
76       1
77       1
78       1
79       1
80       1
81       1
82       1
83       1
84       1
85       1
86       1
87       1
88       1
89       1
90       1

In [69]:
# Count number of occurences of 0 values
print("0 occurence: ",operator.countOf(df_o2_check_2,0))

0 occurence:  105


By summing the categorical o2 variables and printing the occurence of '0', I can see that there are 105 occurences where no categorical o2 variable is marked as positive (1). I now want to change these 105 records to have a value of 1 for the "is_o2_unkwn" variable. My existing "is_o2_unkwn" will hold a value of 1 for all records that have a value of 1 for "is_o2_unkwn" and it will hold a value of 1 for all records that have a value of 0 for "o2_check."

In [70]:
# Create new column 'o2_check' to find records with no value for categorical o2 variables
df_o2_check['o2_check'] = df_o2_check.sum(axis=1)

C:\Users\prena\AppData\Local\Temp\ipykernel_22944\4269333625.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_o2_check['o2_check'] = df_o2_check.sum(axis=1)


In [71]:
df_o2_check.head()

is_o2_not_used  is_o2_climbing  is_o2_descent  is_o2_sleeping  \
0               1               0              0               0   
1               1               0              0               0   
2               1               0              0               0   
3               1               0              0               0   
4               1               0              0               0   

   is_o2_medical  is_o2_used  is_o2_unkwn  had_o2  o2_check  
0              0           0            0       0         1  
1              0           0            0       0         1  
2              0           0            0       0         1  
3              0           0            0       0         1  
4              0           0            0       1         2

In [72]:
# Check frequency for column 'o2_check' PRIOR to updating 'is_o2_unkwn' variable
df_o2_check['o2_check'].value_counts(dropna = False).sort_index()

0     105
1    6516
2    2272
3    1501
4      81
5      14
Name: o2_check, dtype: int64

#### Updating 105 occurences with no categorical O2 data to have a value of one for 'is_o2_unkwn' variable

In [73]:
# Check frequency for column 'is_o2_unkwn' PRIOR to updating 'is_o2_unkwn' variable
df_o2_check['is_o2_unkwn'].value_counts(dropna = False).sort_index()

0    10465
1       24
Name: is_o2_unkwn, dtype: int64

In [74]:
# Updating 'is_o2_unkwn' variable
df_o2_check['is_o2_unkwn'] = np.where(df_o2_check.o2_check == 0,'1',df_o2_check.is_o2_unkwn)

C:\Users\prena\AppData\Local\Temp\ipykernel_22944\1651111791.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_o2_check['is_o2_unkwn'] = np.where(df_o2_check.o2_check == 0,'1',df_o2_check.is_o2_unkwn)


In [75]:
# Check frequency for column 'is_o2_unkwn' AFTER update
df_o2_check['is_o2_unkwn'].value_counts(dropna = False).sort_index()

0    10360
1      129
Name: is_o2_unkwn, dtype: int64

Our 'is_o2_unkwn' variable went from having 24 positive values to 129 positive values (since we changed 105 records).

In [76]:
#Create a dataframe to hold our update 'is_o2_unkwn' variable as a list
new_o2_unkwn = df_o2_check[['is_o2_unkwn']]

In [77]:
new_o2_unkwn.head()

is_o2_unkwn
0           0
1           0
2           0
3           0
4           0

In [78]:
# Check frequency for column 'is_o2_unkwn'
new_o2_unkwn['is_o2_unkwn'].value_counts(dropna = False).sort_index()

0    10360
1      129
Name: is_o2_unkwn, dtype: int64

#### Updating our df_exp_clean_2 dataframe to have the updated values for 'is_o2_unkwn' (to be held in new column 'is_o2_unkwn_2')

In [79]:
# Update our main dataframe to have a new column nameed 'is_o2_unkwn_2' where our updated variable is stored
df_exp_clean_2 = df_exp_clean.assign(is_o2_unkwn_2 = new_o2_unkwn)

In [80]:
df_exp_clean_2.head()

peak_id   peak_name nationality  year  season  bc_arrived     bc_left  \
0    ACHN      Aichyn       Japan  2015  Autumn         NaN         NaN   
1    ACHN      Aichyn      France  2015  Autumn         NaN         NaN   
2    ACHN      Aichyn      France  2018  Autumn         NaN         NaN   
3    AMAD  Ama Dablam          UK  1958  Autumn         NaN         NaN   
4    AMAD  Ama Dablam          UK  1959  Spring  1959-04-17  1959-05-30   

   summit_day  total_days        exp_result  max_elev_reached  total_mbrs  \
0  2015-09-03           0           Success              6055           5   
1  2015-09-25           0           Success              6055           9   
2         NaN           0     Did not Climb                 0           8   
3         NaN           0  Route Difficulty              6000           6   
4  1959-05-21          43          Accident              6550           6   

   mbrs_summited  mbrs_deaths  hired_abc  hired_summits  hired_deaths  \
0              5            0          0              0             0   
1              4            0          2              2             0   
2              0            0          1              0             0   
3              0            0          2              0             0   
4              0            2          7              0             0   

   is_no_hired_abc  is_o2_not_used  is_o2_climbing  is_o2_descent  \
0                1               1               0              0   
1                0               1               0              0   
2                0               1               0              0   
3                0               1               0              0   
4                0               1               0              0   

   is_o2_sleeping  is_o2_medical  is_o2_used  is_o2_unkwn  had_o2  \
0               0              0           0            0       0   
1               0              0           0            0       0   
2               0              0           0            0       0   
3               0              0           0            0       0   
4               0              0           0            0       1   

  is_o2_unkwn_2  
0             0  
1             0  
2             0  
3             0  
4             0

In [81]:
# Check frequency for column 'is_o2_unkwn'
df_exp_clean_2['is_o2_unkwn'].value_counts(dropna = False).sort_index()

0    10465
1       24
Name: is_o2_unkwn, dtype: int64

In [82]:
# Check frequency for column 'is_o2_unkwn_2'
df_exp_clean_2['is_o2_unkwn_2'].value_counts(dropna = False).sort_index()

0    10360
1      129
Name: is_o2_unkwn_2, dtype: int64

Now, we have two columns to represent if oxygen usage on the expedition record is unknown:
- Old variable = 'is_o2_unkwn'
- New variable = 'is_o2_unkwn_2'

In [83]:
# Confirming data type change
df_exp_clean_2['is_o2_unkwn_2'].dtype

dtype('O')

In [84]:
# Changing data type of 'is_o2_unkwn_2' to integer
df_exp_clean_2['is_o2_unkwn_2'] = df_exp_clean_2['is_o2_unkwn_2'].astype('int')

In [85]:
# Confirming data type change
df_exp_clean_2['is_o2_unkwn_2'].dtype

dtype('int32')

#### Creating flag 'o2_check' to check our O2 categorical variables

After changing the data type of our new variable to integer, we can create an 'o2_check' flag for our primary dataframe to add all of the o2 categorical variables.

In [86]:
# Create new column 'o2_check'
df_exp_clean_2['o2_check'] = df_exp_clean_2['is_o2_not_used'] + df_exp_clean_2['is_o2_climbing'] + df_exp_clean_2['is_o2_descent'] + df_exp_clean_2['is_o2_sleeping'] + df_exp_clean_2['is_o2_medical'] + df_exp_clean_2['is_o2_used'] + df_exp_clean_2['had_o2'] + df_exp_clean_2['is_o2_unkwn_2']

In [87]:
df_exp_clean_2.head()

peak_id   peak_name nationality  year  season  bc_arrived     bc_left  \
0    ACHN      Aichyn       Japan  2015  Autumn         NaN         NaN   
1    ACHN      Aichyn      France  2015  Autumn         NaN         NaN   
2    ACHN      Aichyn      France  2018  Autumn         NaN         NaN   
3    AMAD  Ama Dablam          UK  1958  Autumn         NaN         NaN   
4    AMAD  Ama Dablam          UK  1959  Spring  1959-04-17  1959-05-30   

   summit_day  total_days        exp_result  max_elev_reached  total_mbrs  \
0  2015-09-03           0           Success              6055           5   
1  2015-09-25           0           Success              6055           9   
2         NaN           0     Did not Climb                 0           8   
3         NaN           0  Route Difficulty              6000           6   
4  1959-05-21          43          Accident              6550           6   

   mbrs_summited  mbrs_deaths  hired_abc  hired_summits  hired_deaths  \
0              5            0          0              0             0   
1              4            0          2              2             0   
2              0            0          1              0             0   
3              0            0          2              0             0   
4              0            2          7              0             0   

   is_no_hired_abc  is_o2_not_used  is_o2_climbing  is_o2_descent  \
0                1               1               0              0   
1                0               1               0              0   
2                0               1               0              0   
3                0               1               0              0   
4                0               1               0              0   

   is_o2_sleeping  is_o2_medical  is_o2_used  is_o2_unkwn  had_o2  \
0               0              0           0            0       0   
1               0              0           0            0       0   
2               0              0           0            0       0   
3               0              0           0            0       0   
4               0              0           0            0       1   

   is_o2_unkwn_2  o2_check  
0              0         1  
1              0         1  
2              0         1  
3              0         1  
4              0         2

In [88]:
# Check frequency for column 'o2_check'
df_exp_clean_2['o2_check'].value_counts(dropna = False).sort_index()

1    6621
2    2272
3    1501
4      81
5      14
Name: o2_check, dtype: int64

Our 'o2_check' count shows us that there are no more values of 0 for the sum of the 8 categorical o2 variables. One last check before deleting the old 'is_o2_unkwn' variable is to create a subset of our 'is_o2_unkwn_2 = 1' and verify that when filtering for our positive o2 variables, we have a shape of (0,25).

In [89]:
df_is_o2_unkwn_2 = df_exp_clean_2[df_exp_clean_2['is_o2_unkwn_2']==1]

In [90]:
df_is_o2_unkwn_2[df_is_o2_unkwn_2['is_o2_sleeping']==1].shape

(0, 28)

In [91]:
df_is_o2_unkwn_2[df_is_o2_unkwn_2['is_o2_medical']==1].shape

(0, 28)

In [92]:
df_is_o2_unkwn_2[df_is_o2_unkwn_2['is_o2_climbing']==1].shape

(0, 28)

In [93]:
df_is_o2_unkwn_2[df_is_o2_unkwn_2['is_o2_descent']==1].shape

(0, 28)

In [94]:
df_is_o2_unkwn_2[df_is_o2_unkwn_2['is_o2_used']==1].shape

(0, 28)

#### Dropping column 'is_o2_unkwn' since it has been replaced with 'is_o2_unkwn_2'

Last, we can drop our old 'is_o2_unkwn' column:

In [95]:
# Defining new dataframe df_exp_clean_3 (without 'is_o2_unkwn')
df_exp_clean_3 = df_exp_clean_2.drop(columns = ['is_o2_unkwn'])

In [96]:
df_exp_clean_3.shape

(10489, 27)

In [97]:
df_exp_clean_3.head()

peak_id   peak_name nationality  year  season  bc_arrived     bc_left  \
0    ACHN      Aichyn       Japan  2015  Autumn         NaN         NaN   
1    ACHN      Aichyn      France  2015  Autumn         NaN         NaN   
2    ACHN      Aichyn      France  2018  Autumn         NaN         NaN   
3    AMAD  Ama Dablam          UK  1958  Autumn         NaN         NaN   
4    AMAD  Ama Dablam          UK  1959  Spring  1959-04-17  1959-05-30   

   summit_day  total_days        exp_result  max_elev_reached  total_mbrs  \
0  2015-09-03           0           Success              6055           5   
1  2015-09-25           0           Success              6055           9   
2         NaN           0     Did not Climb                 0           8   
3         NaN           0  Route Difficulty              6000           6   
4  1959-05-21          43          Accident              6550           6   

   mbrs_summited  mbrs_deaths  hired_abc  hired_summits  hired_deaths  \
0              5            0          0              0             0   
1              4            0          2              2             0   
2              0            0          1              0             0   
3              0            0          2              0             0   
4              0            2          7              0             0   

   is_no_hired_abc  is_o2_not_used  is_o2_climbing  is_o2_descent  \
0                1               1               0              0   
1                0               1               0              0   
2                0               1               0              0   
3                0               1               0              0   
4                0               1               0              0   

   is_o2_sleeping  is_o2_medical  is_o2_used  had_o2  is_o2_unkwn_2  o2_check  
0               0              0           0       0              0         1  
1               0              0           0       0              0         1  
2               0              0           0       0              0         1  
3               0              0           0       0              0         1  
4               0              0           0       1              0         2

### 3.4 Data Timeliness

There are 10,484 records that fall between 1905 and 2020. Below, I'm calculating how many records there are in the 21st centure (2000-2020), and the most recent 10 years on record (2010-2020). Based on the questions I ask for the project and how many records are available, I could then make a subset for smaller range of years, if needed:

In [98]:
# Checking to see how many records in the 21st century
df_exp_clean_3[df_exp_clean_3['year']>=2000].shape

(6887, 27)

In [99]:
# Checking to see how many records in the most recent 10 years
df_exp_clean_3[df_exp_clean_3['year']>=2010].shape

(3683, 27)

### 3.5 Changing column name "had_o2" to "had_o2_unused"

In [100]:
# Changing column name of 'had_o2' to 'had_o2_unused'
df_exp_clean_3.rename(columns = {'had_o2' : 'had_o2_unused'}, inplace = True)

In [101]:
df_exp_clean_3.head()

peak_id   peak_name nationality  year  season  bc_arrived     bc_left  \
0    ACHN      Aichyn       Japan  2015  Autumn         NaN         NaN   
1    ACHN      Aichyn      France  2015  Autumn         NaN         NaN   
2    ACHN      Aichyn      France  2018  Autumn         NaN         NaN   
3    AMAD  Ama Dablam          UK  1958  Autumn         NaN         NaN   
4    AMAD  Ama Dablam          UK  1959  Spring  1959-04-17  1959-05-30   

   summit_day  total_days        exp_result  max_elev_reached  total_mbrs  \
0  2015-09-03           0           Success              6055           5   
1  2015-09-25           0           Success              6055           9   
2         NaN           0     Did not Climb                 0           8   
3         NaN           0  Route Difficulty              6000           6   
4  1959-05-21          43          Accident              6550           6   

   mbrs_summited  mbrs_deaths  hired_abc  hired_summits  hired_deaths  \
0              5            0          0              0             0   
1              4            0          2              2             0   
2              0            0          1              0             0   
3              0            0          2              0             0   
4              0            2          7              0             0   

   is_no_hired_abc  is_o2_not_used  is_o2_climbing  is_o2_descent  \
0                1               1               0              0   
1                0               1               0              0   
2                0               1               0              0   
3                0               1               0              0   
4                0               1               0              0   

   is_o2_sleeping  is_o2_medical  is_o2_used  had_o2_unused  is_o2_unkwn_2  \
0               0              0           0              0              0   
1               0              0           0              0              0   
2               0              0           0              0              0   
3               0              0           0              0              0   
4               0              0           0              1              0   

   o2_check  
0         1  
1         1  
2         1  
3         1  
4         2

### 3.6 Mixed-Type Columns

In [102]:
# Run code to check for mixed data types
for col in df_exp_clean_3.columns.tolist():
  weird = (df_exp_clean_3[[col]].applymap(type) != df_exp_clean_3[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_exp_clean_3[weird]) > 0:
    print (col)

bc_arrived
bc_left
summit_day


In [103]:
# Checking data types
# Confirming data type change
df_exp_clean_3.dtypes

peak_id             object
peak_name           object
nationality         object
year                 int64
season              object
bc_arrived          object
bc_left             object
summit_day          object
total_days           int64
exp_result          object
max_elev_reached     int64
total_mbrs           int64
mbrs_summited        int64
mbrs_deaths          int64
hired_abc            int64
hired_summits        int64
hired_deaths         int64
is_no_hired_abc      int64
is_o2_not_used       int64
is_o2_climbing       int64
is_o2_descent        int64
is_o2_sleeping       int64
is_o2_medical        int64
is_o2_used           int64
had_o2_unused        int64
is_o2_unkwn_2        int32
o2_check             int64
dtype: object

##### Changing data type for 'bc_arrived' to object

In [104]:
df_exp_clean_3['bc_arrived'].value_counts()

2011-04-14    23
2009-04-19    20
2005-04-14    20
2019-04-15    18
2006-04-10    18
2018-04-16    18
2012-04-15    17
2017-04-17    16
2018-04-18    16
2012-04-16    16
2004-09-08    15
2006-04-15    15
2006-04-05    14
2003-04-07    13
2010-04-09    13
2009-09-10    13
2007-04-07    13
2007-04-12    13
2013-04-12    13
2001-04-05    13
2016-04-19    13
2015-04-15    13
2004-04-16    13
2010-04-14    13
2017-04-16    12
2006-09-11    12
2009-04-15    12
2014-04-21    12
2006-04-07    12
2002-04-11    12
2013-04-15    12
2001-09-02    12
2011-04-17    12
2009-04-12    12
2016-09-11    12
2004-04-05    12
2014-04-11    12
2017-04-18    12
2002-04-12    12
2012-09-11    11
2003-04-08    11
2017-04-14    11
2013-04-18    11
2006-04-09    11
2017-04-20    11
2008-04-12    11
2014-04-17    11
2019-04-17    11
2012-04-17    11
2010-09-09    11
2010-04-11    11
2011-04-13    11
2010-09-10    11
2010-04-07    10
2002-04-08    10
2004-04-08    10
2013-04-10    10
2019-04-13    10
2005-04-05    

In [105]:
# Changing data type of column 'bc_arrived' to object
df_exp_clean_3['bc_arrived'] = df_exp_clean_3['bc_arrived'].astype('object')

In [106]:
df_exp_clean_3['bc_arrived'].value_counts()

2011-04-14    23
2009-04-19    20
2005-04-14    20
2019-04-15    18
2006-04-10    18
2018-04-16    18
2012-04-15    17
2017-04-17    16
2018-04-18    16
2012-04-16    16
2004-09-08    15
2006-04-15    15
2006-04-05    14
2003-04-07    13
2010-04-09    13
2009-09-10    13
2007-04-07    13
2007-04-12    13
2013-04-12    13
2001-04-05    13
2016-04-19    13
2015-04-15    13
2004-04-16    13
2010-04-14    13
2017-04-16    12
2006-09-11    12
2009-04-15    12
2014-04-21    12
2006-04-07    12
2002-04-11    12
2013-04-15    12
2001-09-02    12
2011-04-17    12
2009-04-12    12
2016-09-11    12
2004-04-05    12
2014-04-11    12
2017-04-18    12
2002-04-12    12
2012-09-11    11
2003-04-08    11
2017-04-14    11
2013-04-18    11
2006-04-09    11
2017-04-20    11
2008-04-12    11
2014-04-17    11
2019-04-17    11
2012-04-17    11
2010-09-09    11
2010-04-11    11
2011-04-13    11
2010-09-10    11
2010-04-07    10
2002-04-08    10
2004-04-08    10
2013-04-10    10
2019-04-13    10
2005-04-05    

##### Changing data type for 'bc_left' to object

In [107]:
df_exp_clean_3['bc_left'].value_counts()

2006-05-21    26
2009-05-23    24
2019-05-25    23
2011-05-24    23
2013-05-24    23
2012-05-28    22
2013-05-23    21
2010-05-26    21
2013-05-25    20
2009-09-30    20
2018-05-21    20
2010-05-27    20
2003-05-25    19
2014-05-21    19
2017-05-23    19
2013-05-26    19
2009-05-24    18
2006-10-05    18
2011-10-06    17
2012-05-23    17
2000-05-25    16
2016-05-22    16
2012-05-22    16
2008-05-25    16
2011-05-23    15
2013-05-21    15
2016-05-24    15
2017-05-29    15
2003-05-27    15
2017-05-28    15
2019-05-26    15
2015-05-01    15
2006-05-26    15
2014-10-04    15
2019-09-29    15
2005-06-02    15
2012-05-13    15
2008-05-26    15
2010-05-25    14
2013-10-05    14
2007-10-05    14
2006-10-04    14
2006-05-23    14
2004-05-26    14
2010-05-28    14
2011-05-25    14
2007-05-21    14
2011-05-22    13
2009-05-25    13
2007-05-26    13
2019-05-27    13
2006-05-19    13
2006-05-20    13
2018-05-22    13
2001-05-27    13
2018-05-20    13
2013-05-22    13
2016-10-02    13
2008-05-27    

In [108]:
# Changing data type of column 'bc_left' to object
df_exp_clean_3['bc_left'] = df_exp_clean_3['bc_left'].astype('object')

In [109]:
df_exp_clean_3['bc_left'].value_counts()

2006-05-21    26
2009-05-23    24
2019-05-25    23
2011-05-24    23
2013-05-24    23
2012-05-28    22
2013-05-23    21
2010-05-26    21
2013-05-25    20
2009-09-30    20
2018-05-21    20
2010-05-27    20
2003-05-25    19
2014-05-21    19
2017-05-23    19
2013-05-26    19
2009-05-24    18
2006-10-05    18
2011-10-06    17
2012-05-23    17
2000-05-25    16
2016-05-22    16
2012-05-22    16
2008-05-25    16
2011-05-23    15
2013-05-21    15
2016-05-24    15
2017-05-29    15
2003-05-27    15
2017-05-28    15
2019-05-26    15
2015-05-01    15
2006-05-26    15
2014-10-04    15
2019-09-29    15
2005-06-02    15
2012-05-13    15
2008-05-26    15
2010-05-25    14
2013-10-05    14
2007-10-05    14
2006-10-04    14
2006-05-23    14
2004-05-26    14
2010-05-28    14
2011-05-25    14
2007-05-21    14
2011-05-22    13
2009-05-25    13
2007-05-26    13
2019-05-27    13
2006-05-19    13
2006-05-20    13
2018-05-22    13
2001-05-27    13
2018-05-20    13
2013-05-22    13
2016-10-02    13
2008-05-27    

##### Changing data type for 'summit_day' to object

In [110]:
df_exp_clean_3['summit_day'].value_counts()

2012-05-19    53
2019-05-23    38
2016-05-19    36
2011-05-20    36
2009-05-19    33
2010-05-23    33
2015-04-25    32
2003-05-22    30
2011-05-21    30
2006-10-01    29
2006-05-17    28
2016-09-30    28
2013-05-19    27
2014-05-18    27
2010-05-17    26
2008-05-21    26
2002-05-16    26
2019-05-22    26
2013-05-20    25
2006-05-18    25
2012-05-25    24
2001-05-23    24
2017-05-21    23
2005-05-30    22
2018-05-16    22
2013-05-21    22
2013-05-22    22
2011-10-04    22
2009-05-20    21
2007-05-21    21
2018-05-18    21
2007-05-16    21
2019-05-16    20
2013-10-02    20
2018-09-28    20
2010-05-24    19
2009-05-18    19
2009-09-24    19
2006-10-02    18
2017-05-16    18
2005-05-21    18
2018-05-19    18
2004-05-16    18
2019-09-26    17
2017-05-20    17
2009-05-21    17
2002-05-17    16
2010-10-01    16
2014-05-25    16
2003-05-20    16
2014-04-18    15
2017-05-22    15
2007-05-15    15
2008-05-22    14
2004-05-18    14
2001-05-14    14
2011-05-19    14
2008-10-03    13
2012-05-18    

In [111]:
# Changing data type of column 'summit_day' to object
df_exp_clean_3['summit_day'] = df_exp_clean_3['summit_day'].astype('object')

In [112]:
df_exp_clean_3['summit_day'].value_counts()

2012-05-19    53
2019-05-23    38
2016-05-19    36
2011-05-20    36
2009-05-19    33
2010-05-23    33
2015-04-25    32
2003-05-22    30
2011-05-21    30
2006-10-01    29
2006-05-17    28
2016-09-30    28
2013-05-19    27
2014-05-18    27
2010-05-17    26
2008-05-21    26
2002-05-16    26
2019-05-22    26
2013-05-20    25
2006-05-18    25
2012-05-25    24
2001-05-23    24
2017-05-21    23
2005-05-30    22
2018-05-16    22
2013-05-21    22
2013-05-22    22
2011-10-04    22
2009-05-20    21
2007-05-21    21
2018-05-18    21
2007-05-16    21
2019-05-16    20
2013-10-02    20
2018-09-28    20
2010-05-24    19
2009-05-18    19
2009-09-24    19
2006-10-02    18
2017-05-16    18
2005-05-21    18
2018-05-19    18
2004-05-16    18
2019-09-26    17
2017-05-20    17
2009-05-21    17
2002-05-17    16
2010-10-01    16
2014-05-25    16
2003-05-20    16
2014-04-18    15
2017-05-22    15
2007-05-15    15
2008-05-22    14
2004-05-18    14
2001-05-14    14
2011-05-19    14
2008-10-03    13
2012-05-18    

#### Check for mixed data types round II

In [113]:
# Run code to check for mixed data types
for col in df_exp_clean_3.columns.tolist():
  weird = (df_exp_clean_3[[col]].applymap(type) != df_exp_clean_3[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_exp_clean_3[weird]) > 0:
    print (col)

bc_arrived
bc_left
summit_day


### 3.7 Outliers

In [114]:
df_exp_clean_3.describe()[['total_days','max_elev_reached','total_mbrs','mbrs_summited','mbrs_deaths','hired_abc','hired_summits','hired_deaths']]

total_days  max_elev_reached    total_mbrs  mbrs_summited  \
count  10489.000000      10489.000000  10489.000000   10489.000000   
mean      20.280008       7107.830870      5.944704       1.867576   
std       17.654847       1761.100193      5.411375       2.895047   
min        0.000000          0.000000      0.000000       0.000000   
25%        4.000000       6530.000000      2.000000       0.000000   
50%       18.000000       7200.000000      5.000000       1.000000   
75%       32.000000       8188.000000      8.000000       3.000000   
max      280.000000       8850.000000     99.000000      39.000000   

        mbrs_deaths     hired_abc  hired_summits  hired_deaths  
count  10489.000000  10489.000000   10489.000000  10489.000000  
mean       0.074840      2.787968       1.010106      0.030317  
std        0.374921      5.067211       2.437881      0.283065  
min        0.000000      0.000000       0.000000      0.000000  
25%        0.000000      0.000000       0.000000      0.000000  
50%        0.000000      1.000000       0.000000      0.000000  
75%        0.000000      3.000000       1.000000      0.000000  
max       10.000000     99.000000      48.000000     11.000000

#### Removing records with value of 0 for max_elev_reached, total_mbrs, and total_days

In [115]:
# Seeing how man max_elev_reached values = 0
df_exp_clean_3[(df_exp_clean_3.max_elev_reached == 0)].shape

(423, 27)

In [116]:
# Creating subset with max_elev_reached > 0
df_exp_clean_4 = df_exp_clean_3[(df_exp_clean_3.max_elev_reached > 0)]

In [117]:
df_exp_clean_4.shape

(10066, 27)

In [118]:
# Seeing how many total_days values = 0
df_exp_clean_4[(df_exp_clean_4.total_days == 0)].shape

(2084, 27)

In [119]:
# Creating subset without total_days > 0
df_exp_clean_5 = df_exp_clean_4[(df_exp_clean_4.total_days > 0)]

In [120]:
df_exp_clean_5.shape

(7982, 27)

In [121]:
# Seeing how many total_mbrs == 0
df_exp_clean_5[(df_exp_clean_5.total_mbrs == 0)].shape

(0, 27)

By excluding zero values from the following fields:
- total_days
- max_elev_reached
- total_mbrs

#### Our resulting dataset has 7982 records

In [122]:
df_exp_clean_5.describe()

year   total_days  max_elev_reached   total_mbrs  mbrs_summited  \
count  7982.000000  7982.000000       7982.000000  7982.000000    7982.000000   
mean   2003.619394    26.552117       7451.066775     5.798797       1.873089   
std      11.849882    15.573310       1009.599699     5.076570       2.843591   
min    1921.000000     1.000000       3800.000000     1.000000       0.000000   
25%    1998.000000    14.000000       6800.000000     2.000000       0.000000   
50%    2006.000000    25.000000       7400.000000     4.000000       1.000000   
75%    2012.000000    37.000000       8188.000000     8.000000       3.000000   
max    2020.000000   280.000000       8850.000000    99.000000      32.000000   

       mbrs_deaths    hired_abc  hired_summits  hired_deaths  is_no_hired_abc  \
count  7982.000000  7982.000000    7982.000000   7982.000000      7982.000000   
mean      0.061639     2.770358       1.061889      0.027186         0.325232   
std       0.323586     4.697144       2.419158      0.237743         0.468491   
min       0.000000     0.000000       0.000000      0.000000         0.000000   
25%       0.000000     0.000000       0.000000      0.000000         0.000000   
50%       0.000000     2.000000       0.000000      0.000000         0.000000   
75%       0.000000     3.000000       1.000000      0.000000         1.000000   
max       5.000000    99.000000      43.000000      7.000000         1.000000   

       is_o2_not_used  is_o2_climbing  is_o2_descent  is_o2_sleeping  \
count     7982.000000     7982.000000    7982.000000     7982.000000   
mean         0.681533        0.278752       0.015535        0.168379   
std          0.465911        0.448413       0.123675        0.374226   
min          0.000000        0.000000       0.000000        0.000000   
25%          0.000000        0.000000       0.000000        0.000000   
50%          1.000000        0.000000       0.000000        0.000000   
75%          1.000000        1.000000       0.000000        0.000000   
max          1.000000        1.000000       1.000000        1.000000   

       is_o2_medical   is_o2_used  had_o2_unused  is_o2_unkwn_2     o2_check  
count    7982.000000  7982.000000    7982.000000    7982.000000  7982.000000  
mean        0.034453     0.314959       0.100476       0.004385     1.598472  
std         0.182400     0.464529       0.300653       0.066077     0.797039  
min         0.000000     0.000000       0.000000       0.000000     1.000000  
25%         0.000000     0.000000       0.000000       0.000000     1.000000  
50%         0.000000     0.000000       0.000000       0.000000     1.000000  
75%         0.000000     1.000000       0.000000       0.000000     2.000000  
max         1.000000     1.000000       1.000000       1.000000     5.000000

#### Defining number of outliers for max_elev_reached, total_mbrs, and total_days

In [123]:
#create a function to find outliers using IQR
def find_outliers_IQR(df_exp_clean_5):
   q1=df_exp_clean_5.quantile(0.25)
   q3=df_exp_clean_5.quantile(0.75)
   IQR=q3-q1
   outliers = df_exp_clean_5[((df_exp_clean_5<(q1-1.5*IQR)) | (df_exp_clean_5>(q3+1.5*IQR)))]
   return outliers

In [124]:
df_exp_clean_5.quantile(0.25)

year                1998.0
total_days            14.0
max_elev_reached    6800.0
total_mbrs             2.0
mbrs_summited          0.0
mbrs_deaths            0.0
hired_abc              0.0
hired_summits          0.0
hired_deaths           0.0
is_no_hired_abc        0.0
is_o2_not_used         0.0
is_o2_climbing         0.0
is_o2_descent          0.0
is_o2_sleeping         0.0
is_o2_medical          0.0
is_o2_used             0.0
had_o2_unused          0.0
is_o2_unkwn_2          0.0
o2_check               1.0
Name: 0.25, dtype: float64

In [125]:
q1 = df_exp_clean_5.quantile(0.25)

In [126]:
df_exp_clean_5.quantile(0.75)

year                2012.0
total_days            37.0
max_elev_reached    8188.0
total_mbrs             8.0
mbrs_summited          3.0
mbrs_deaths            0.0
hired_abc              3.0
hired_summits          1.0
hired_deaths           0.0
is_no_hired_abc        1.0
is_o2_not_used         1.0
is_o2_climbing         1.0
is_o2_descent          0.0
is_o2_sleeping         0.0
is_o2_medical          0.0
is_o2_used             1.0
had_o2_unused          0.0
is_o2_unkwn_2          0.0
o2_check               2.0
Name: 0.75, dtype: float64

In [127]:
q3 = df_exp_clean_5.quantile(0.75)

In [128]:
q3 - q1

year                  14.0
total_days            23.0
max_elev_reached    1388.0
total_mbrs             6.0
mbrs_summited          3.0
mbrs_deaths            0.0
hired_abc              3.0
hired_summits          1.0
hired_deaths           0.0
is_no_hired_abc        1.0
is_o2_not_used         1.0
is_o2_climbing         1.0
is_o2_descent          0.0
is_o2_sleeping         0.0
is_o2_medical          0.0
is_o2_used             1.0
had_o2_unused          0.0
is_o2_unkwn_2          0.0
o2_check               1.0
dtype: float64

In [129]:
IQR = q3 - q1

#### Variable 'total_days' Outliers

In [130]:
outliers_total_days = find_outliers_IQR(df_exp_clean_5['total_days'])

In [131]:
print('number of outliers: '+ str(len(outliers_total_days)))

number of outliers: 50


In [132]:
q1.total_days-1.5*IQR.total_days

-20.5

In [133]:
q3.total_days+1.5*IQR.total_days

71.5

In [134]:
df_exp_clean_5[(df_exp_clean_5.total_days > -20.5) & (df_exp_clean_5.total_days <= 71.5)].shape

(7932, 27)

In [135]:
#Check dataset size minus count of outliers = new dataset size
7978 - 50

7928

#### Variable 'max_elev_reached' Outliers

In [136]:
outliers_max_elev = find_outliers_IQR(df_exp_clean_5['max_elev_reached'])

In [137]:
print('number of outliers: '+ str(len(outliers_max_elev)))

number of outliers: 15


In [138]:
q1.max_elev_reached-1.5*IQR.max_elev_reached

4718.0

In [139]:
q3.max_elev_reached+1.5*IQR.max_elev_reached

10270.0

In [140]:
df_exp_clean_5[(df_exp_clean_5.max_elev_reached >= 4718) & (df_exp_clean_5.max_elev_reached <= 10270)].shape

(7967, 27)

In [141]:
# Check dataset size minuse count of outliers = new dataset size
7978 - 15

7963

#### Variable 'total_mbrs' Outliers

In [142]:
outliers_total_mbrs = find_outliers_IQR(df_exp_clean_5['total_mbrs'])

In [143]:
print('number of outliers: '+ str(len(outliers_total_mbrs)))

number of outliers: 231


In [144]:
q1.total_mbrs-1.5*IQR.total_mbrs

-7.0

In [145]:
q3.total_mbrs+1.5*IQR.total_mbrs

17.0

In [146]:
df_exp_clean_5[(df_exp_clean_5.total_mbrs >= -7) & (df_exp_clean_5.total_mbrs <= 17)].shape

(7751, 27)

In [147]:
# Check dataset size minuse count of outliers = new dataset size
7978 - 231

7747

## Section 4: Exporting dataframe "df_exp_clean_3" as "expeditions_clean.pkl"

In [148]:
df_exp_clean_5.shape

(7982, 27)

In [149]:
# Export data to pkl
df_exp_clean_5.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'expeditions_clean.pkl'))